mock/options.txt
version:512
invertYMouse:false
mouseSensitivity:0.47887325
fov:0.0
gamma:1.0
saturation:0.0
renderDistance:12
guiScale:0
particles:1
bobView:true
anaglyph3d:false
maxFps:120
fboEnable:true
difficulty:1
fancyGraphics:false
ao:1
renderClouds:false
resourcePacks:["Xray Ultimate 1.12 v2.2.1.zip"]
incompatibleResourcePacks:[]
lastServer:play.mcndsj.com
lang:en_US
chatVisibility:0
chatColors:true
chatLinks:true
chatLinksPrompt:true
chatOpacity:1.0
snooperEnabled:true
fullscreen:false
enableVsync:true
useVbo:true
hideServerAddress:false
advancedItemTooltips:false
pauseOnLostFocus:true
touchscreen:false
overrideWidth:0
overrideHeight:0
heldItemTooltips:true
chatHeightFocused:1.0
chatHeightUnfocused:0.44366196
chatScale:1.0
chatWidth:1.0
showInventoryAchievementHint:false
mipmapLevels:4
forceUnicodeFont:false
reducedDebugInfo:false
useNativeTransport:true
entityShadows:true
mainHand:right
attackIndicator:1
showSubtitles:false
realmsNotifications:true
enableWeakAttacks:false
autoJump:true
key_key.attack:-100
key_key.use:-99
key_key.forward:17
key_key.left:30
key_key.back:31
key_key.right:32
key_key.jump:57
key_key.sneak:42
key_key.sprint:29
key_key.drop:16
key_key.inventory:18
key_key.chat:28
key_key.playerlist:15
key_key.pickItem:-98
key_key.command:53
key_key.screenshot:60
key_key.togglePerspective:63
key_key.smoothCamera:0
key_key.fullscreen:87
key_key.spectatorOutlines:0
key_key.swapHands:33
key_key.hotbar.1:2
key_key.hotbar.2:3
key_key.hotbar.3:4
key_key.hotbar.4:5
key_key.hotbar.5:6
key_key.hotbar.6:7
key_key.hotbar.7:8
key_key.hotbar.8:9
key_key.hotbar.9:10
soundCategory_master:1.0
soundCategory_music:1.0
soundCategory_record:1.0
soundCategory_weather:1.0
soundCategory_block:1.0
soundCategory_hostile:1.0
soundCategory_neutral:1.0
soundCategory_player:1.0
soundCategory_ambient:1.0
soundCategory_voice:1.0
modelPart_cape:true
modelPart_jacket:true
modelPart_left_sleeve:true
modelPart_right_sleeve:true
modelPart_left_pants_leg:true
modelPart_right_pants_leg:true
modelPart_hat:true


CHANGELOG.md

## 0.17.1
### xmcl@0.16.2
#### Bug Fixes

- auto update for nsis ([db668ca243c9f2b363f227bf35e300ce54ae980d](https://github.com/Voxelum/x-minecraft-launcher/commit/db668ca243c9f2b363f227bf35e300ce54ae980d))
### @xmcl/keystone-ui@0.16.1
#### Bug Fixes

- **ui**: correctly detect the loading state ([8bcb1790439d7b6b4f3edb0f9b07469dd48c3717](https://github.com/Voxelum/x-minecraft-launcher/commit/8bcb1790439d7b6b4f3edb0f9b07469dd48c3717))
- disable update if it's in appx container ([3303c05e830c43f78a1fd96e0baef80a5a2594d4](https://github.com/Voxelum/x-minecraft-launcher/commit/3303c05e830c43f78a1fd96e0baef80a5a2594d4))
 
### @xmcl/runtime@0.15.1
#### Bug Fixes

- disable update if it's in appx container ([3303c05e830c43f78a1fd96e0baef80a5a2594d4](https://github.com/Voxelum/x-minecraft-launcher/commit/3303c05e830c43f78a1fd96e0baef80a5a2594d4))

## 0.17.0
### xmcl@0.16.1
#### Bug Fixes

- remove dead code & update schema ([d20b86767ceb879ce170f6ccec7d33570013b134](https://github.com/Voxelum/x-minecraft-launcher/commit/d20b86767ceb879ce170f6ccec7d33570013b134))
- initial lang & lang update on tray menu ([5cc7a0b9e18486415ae8a3762b00d8b07a5da64c](https://github.com/Voxelum/x-minecraft-launcher/commit/5cc7a0b9e18486415ae8a3762b00d8b07a5da64c))
### @xmcl/runtime-api@0.1.0
#### Features

- support modrinth ([a903f8f35da14ecb9841c2c5dff2d70d0c10b758](https://github.com/Voxelum/x-minecraft-launcher/commit/a903f8f35da14ecb9841c2c5dff2d70d0c10b758))
#### Bug Fixes

- mod import to instance resolved to wrong path ([8d5ed52b371f3407b0fc235fa1a9b0e7a88eaf00](https://github.com/Voxelum/x-minecraft-launcher/commit/8d5ed52b371f3407b0fc235fa1a9b0e7a88eaf00))
- remove dead code & update schema ([d20b86767ceb879ce170f6ccec7d33570013b134](https://github.com/Voxelum/x-minecraft-launcher/commit/d20b86767ceb879ce170f6ccec7d33570013b134))
- initial lang & lang update on tray menu ([5cc7a0b9e18486415ae8a3762b00d8b07a5da64c](https://github.com/Voxelum/x-minecraft-launcher/commit/5cc7a0b9e18486415ae8a3762b00d8b07a5da64c))
- correctly handle default java path ([4f2d04e42fd12dceb3a9072c3d72d84cdab41fcd](https://github.com/Voxelum/x-minecraft-launcher/commit/4f2d04e42fd12dceb3a9072c3d72d84cdab41fcd))
### @xmcl/keystone-ui@0.16.0
#### Features

- **ui**: use halo as default ([2833d21e34f55930ce815276d5db7ed76aa07ea4](https://github.com/Voxelum/x-minecraft-launcher/commit/2833d21e34f55930ce815276d5db7ed76aa07ea4))
- support modrinth ([a903f8f35da14ecb9841c2c5dff2d70d0c10b758](https://github.com/Voxelum/x-minecraft-launcher/commit/a903f8f35da14ecb9841c2c5dff2d70d0c10b758))
#### Bug Fixes

- import modpack dialog non-reactive after an import ([8b687bc34e4242c02a70d49fca11e5ce3998e142](https://github.com/Voxelum/x-minecraft-launcher/commit/8b687bc34e4242c02a70d49fca11e5ce3998e142))
- remove dead code & update schema ([d20b86767ceb879ce170f6ccec7d33570013b134](https://github.com/Voxelum/x-minecraft-launcher/commit/d20b86767ceb879ce170f6ccec7d33570013b134))
- should force user to login before launch ([d7603db5e8e07d99077065bbe49d1c53643dfd83](https://github.com/Voxelum/x-minecraft-launcher/commit/d7603db5e8e07d99077065bbe49d1c53643dfd83))
- active curseforge icon when route is match ([ef74858172b96f0f9541a9ad45966cf1a11a3aa9](https://github.com/Voxelum/x-minecraft-launcher/commit/ef74858172b96f0f9541a9ad45966cf1a11a3aa9))
- the bar is unclickable when there is no item ([7732204d054bbd18be7c40f30ab7b372ae312551](https://github.com/Voxelum/x-minecraft-launcher/commit/7732204d054bbd18be7c40f30ab7b372ae312551))
- allow to download java if the local java is invalid ([0073070ee00b03268aa1d46a8bf40b5c65c6c8a2](https://github.com/Voxelum/x-minecraft-launcher/commit/0073070ee00b03268aa1d46a8bf40b5c65c6c8a2))
- correclty sort the tasks ([36143fab4136cfbf953763218144bc88551b8bad](https://github.com/Voxelum/x-minecraft-launcher/commit/36143fab4136cfbf953763218144bc88551b8bad))
- quit launcher once click the close btn ([24c2e510bebfcc29fa34e824565168efc624bf2d](https://github.com/Voxelum/x-minecraft-launcher/commit/24c2e510bebfcc29fa34e824565168efc624bf2d))
#### Dependencies Updates

### @xmcl/runtime@0.15.0
#### Features

- support modrinth ([a903f8f35da14ecb9841c2c5dff2d70d0c10b758](https://github.com/Voxelum/x-minecraft-launcher/commit/a903f8f35da14ecb9841c2c5dff2d70d0c10b758))
#### Bug Fixes

- mod import to instance resolved to wrong path ([8d5ed52b371f3407b0fc235fa1a9b0e7a88eaf00](https://github.com/Voxelum/x-minecraft-launcher/commit/8d5ed52b371f3407b0fc235fa1a9b0e7a88eaf00))
- remove dead code & update schema ([d20b86767ceb879ce170f6ccec7d33570013b134](https://github.com/Voxelum/x-minecraft-launcher/commit/d20b86767ceb879ce170f6ccec7d33570013b134))
- correctly handle default java path ([4f2d04e42fd12dceb3a9072c3d72d84cdab41fcd](https://github.com/Voxelum/x-minecraft-launcher/commit/4f2d04e42fd12dceb3a9072c3d72d84cdab41fcd))
- crash on diagnose version ([756167c654789ee63213a14459734694f1bbd607](https://github.com/Voxelum/x-minecraft-launcher/commit/756167c654789ee63213a14459734694f1bbd607))
#### Dependencies Updates



## 0.16.0
### xmcl@0.16.0
#### Features

- support the app top control bar ([6499ccdd907f1b8ae7a75032b020e97602258117](https://github.com/Voxelum/x-minecraft-launcher/commit/6499ccdd907f1b8ae7a75032b020e97602258117))
### @xmcl/runtime-api@0.0.2
#### Bug Fixes

- use new pipeline ([480df463c0822789639297d097057a024fd68351](https://github.com/Voxelum/x-minecraft-launcher/commit/480df463c0822789639297d097057a024fd68351))
### @xmcl/keystone-ui@0.15.0
#### Features

- support the app top control bar ([6499ccdd907f1b8ae7a75032b020e97602258117](https://github.com/Voxelum/x-minecraft-launcher/commit/6499ccdd907f1b8ae7a75032b020e97602258117))
#### Bug Fixes

- add some missing locale ([9d1789eb268fc8abed516ac6bbda1f8e076e186c](https://github.com/Voxelum/x-minecraft-launcher/commit/9d1789eb268fc8abed516ac6bbda1f8e076e186c))
#### Dependencies Updates

### @xmcl/runtime@0.14.1
#### Dependencies Updates




src/renderer/store/modules/profiles/modules/settings.js
import { GameSetting, WorldInfo } from 'ts-minecraft'
import Vue from 'vue'
import fs from 'fs-extra'
import paths from 'path'
import Zip from 'jszip'

export default {
    state: () => ({
        version: 1139, // for 1.12
        invertYMouse: false,
        mouseSensitivity: 0.5,
        difficulty: 2,

        // critical performance video settings 
        renderDistance: 12,
        particles: 0,
        fboEnable: true,
        fancyGraphics: true,
        ao: 2,
        renderClouds: true,
        enableVsync: true,
        useVbo: true,
        mipmapLevels: 4,
        anaglyph3d: false,

        fov: 0,
        gamma: 0,
        saturation: 0,
        guiScale: 0,
        bobView: true,
        maxFps: 120,
        fullscreen: false,

        resourcePacks: [],
        incompatibleResourcePacks: [],
        lastServer: '',
        lang: 'en_us',
        chatVisibility: 0,
        chatColors: true,
        chatLinks: true,
        chatLinksPrompt: true,
        chatOpacity: 1,
        snooperEnabled: true,

        hideServerAddress: false,
        advancedItemTooltips: false,
        pauseOnLostFocus: true,
        touchscreen: false,
        overrideWidth: 0,
        overrideHeight: 0,
        heldItemTooltips: true,
        chatHeightFocused: 1,
        chatHeightUnfocused: 0.44366196,
        chatScale: 1,
        chatWidth: 1,
        forceUnicodeFont: false,
        reducedDebugInfo: false,
        useNativeTransport: true,
        entityShadows: true,
        mainHand: 'right',
        attackIndicator: 1,
        showSubtitles: false,
        realmsNotifications: true,
        enableWeakAttacks: false,
        autoJump: true,
        narrator: 0,
        tutorialStep: 'movement',
        'key_key.attack': -100,
        'key_key.use': -99,
        'key_key.forward': 17,
        'key_key.left': 30,
        'key_key.back': 31,
        'key_key.right': 32,
        'key_key.jump': 57,
        'key_key.sneak': 42,
        'key_key.sprint': 29,
        'key_key.drop': 16,
        'key_key.inventory': 18,
        'key_key.chat': 20,
        'key_key.playerlist': 15,
        'key_key.pickItem': -98,
        'key_key.command': 53,
        'key_key.screenshot': 60,
        'key_key.togglePerspective': 63,
        'key_key.smoothCamera': 0,
        'key_key.fullscreen': 87,
        'key_key.spectatorOutlines': 0,
        'key_key.swapHands': 33,
        'key_key.saveToolbarActivator': 46,
        'key_key.loadToolbarActivator': 45,
        'key_key.advancements': 38,
        'key_key.hotbar.1': 2,
        'key_key.hotbar.2': 3,
        'key_key.hotbar.3': 4,
        'key_key.hotbar.4': 5,
        'key_key.hotbar.5': 6,
        'key_key.hotbar.6': 7,
        'key_key.hotbar.7': 8,
        'key_key.hotbar.8': 9,
        'key_key.hotbar.9': 10,
        soundCategory_master: 1,
        soundCategory_music: 1,
        soundCategory_record: 1,
        soundCategory_weather: 1,
        soundCategory_block: 1,
        soundCategory_hostile: 1,
        soundCategory_neutral: 1,
        soundCategory_player: 1,
        soundCategory_ambient: 1,
        soundCategory_voice: 1,
        modelPart_cape: true,
        modelPart_jacket: true,
        modelPart_left_sleeve: true,
        modelPart_right_sleeve: true,
        modelPart_left_pants_leg: true,
        modelPart_right_pants_leg: true,
        modelPart_hat: true,
    }),
    getters: {
        mcoptions: state => state,
        resourcepacks: state => state.resourcePacks,
        mclanguage: state => state.lang || '',
    },
    mutations: {
        mcoption(states, { key, value }) {
            states[key] = value
        },
        mcoptions(state, settings) {
            Object.keys(settings)
                .filter(k => state[k] !== undefined)
                .forEach((k) => {
                    if (settings[k] !== null) {
                        if (state[k] instanceof Array) {
                            if (settings[k] instanceof Array) {
                                state[k] = settings[k].slice()
                            } else {
                                console.error(`Cannot copy the value ${settings[k]}`)
                                console.error(settings);
                            }
                        } else {
                            state[k] = settings[k]
                        }
                    }
                })
        },

        /**
         * 
         * @param {any} states 
         * @param {{action:ResourcePackActions, pack:string}} payload 
         */
        resourcepack(states, payload) {
            const { action, pack } = payload;
            let idx;
            let temp;
            switch (action) {
                case 'add':
                    if (states.resourcePacks.indexOf(pack) === -1) {
                        states.resourcePacks.push(pack)
                    }
                    break;
                case 'remove':
                    Vue.delete(states.resourcePacks, states.resourcePacks.indexOf(pack));
                    break;
                case 'moveup':
                    idx = states.resourcePacks.indexOf(pack)
                    if (idx <= 0) return;
                    temp = states.resourcePacks[idx - 1];
                    Vue.set(states.resourcePacks, idx - 1, pack)
                    Vue.set(states.resourcePacks, idx, temp)
                    break;
                case 'movedown':
                    idx = states.resourcePacks.indexOf(pack)
                    if (idx === -1 || idx === states.resourcePacks.length - 1) return;
                    temp = states.resourcePacks[idx + 1];
                    Vue.set(states.resourcePacks, idx + 1, pack)
                    Vue.set(states.resourcePacks, idx, temp)
                    break;
                default: break;
            }
        },
    },
    actions: {
        saveSetting(context) {
            const path = `profiles/${context.state.id}/options.txt`
            const snapshot = JSON.stringify(context.state);
            const data = GameSetting.stringify(context.state);
            return context.dispatch('write', { path, data }, { root: true })
        },
        async load(context, { id }) {
            const gcString = await context.dispatch('read', {
                path: `profiles/${id}/options.txt`,
                type: 'string',
            }, { root: true });
            context.commit('mcoptions', GameSetting.parseFrame(gcString));
            context.dispatch('saveSetting');
        },
    },
}
